ReadMe.md
# Reconhecimento Facial para a Classificação e Registro de presença em sala de aula

Este repositório é o Trabalho de Conclusão de Curso
desenvolvido por Alleff Dymytry Pereira de Deus como
requisito para a graduação em Bacharel de Engenharia de computação na Universidade Estadual do Rio Grande do Sul.

Esse repositório foi criado também com o intuito de obter uma melhor organização do versionamento e controle dos códigos e documentação atrelados ao TCC.

*Autor:* **Alleff Dymytry**
*Orientadora:* **Letícia Vieira Guimarães**

changelog.md
# Lista TO DO do TCC
 
    1. Organizar funções na Tela_principal.py para realizar os testes dos resultados necesários no documento.
    2. Fazer o cálculo dos pesos a partir da face média previamente obtida com a face cropada do HaarLike.py.
    3. Receber da Tela_principal.py um vetor(tamanho 10) com a face já selecionada.
    4. Escrever sobre as modificações feitas na Tela_principal.py no TCC.
    5. Colocar a matriz de covariância no drive, para a função get_omegaface() do arquivo Eigen.py possa utilizar a partir da Tela_principal.py.

# Changelog

## 11/12/2020

    1.  Modificação na Documentação em relação as mudanças sugeridas pela orientadora.

## 10/12/2020

    1.  Modificação no arquivo Eigen.py para calcular o tempo que leva para buscar os arquivos do drive;
    2.  Criação de funções para obter os arquivos do drive a partir de seus ids;
    3.  Modificação dos arquivos Database.py e Google_drive.py para que a função de busca com id pudesse funcionar;
    4.  Refatoração do Banco do Mongo, onde os arquivos foram separados no seguinte esquema: {Images,Cov e Mean};
    5.  Reinserção dos arquivos no Drive;

## 07/12/2020

    1.  Modificação no arquivo Google_Drive.py para que possa pegar a face Psi diretamente do Drive;
    2.  Modificação no arquivo Eigen.py para receber a face Psi do arquivo Google_Drive.py;

## 01/12/2020

    1.  Modificações na Documentação
    2.  Criados novos meios de reconhecimento automático na função Eigen.py
    3.  Tarefas feitas do TO DO:
        1.  Pegar as imagens a partir do Drive para a função do Eigenface;
        2.  Criar a função que lê o JSON da imagem e transforma no vetor Gamma;
        3.  Escrever no texto o que já foi feito até o momento e estruturar os materiais e métodos utilizados referenciando os autores necessários;
        4.  Fazer os cálculos euclidianos das faces perante as faces de fora do treinamento.

## 07/11/2020

    1.  Modificação na Documentação.
    2.  Modificações na função do arquivo Eigen.py para salvar as imagens para a documentação.

## 06/11/2020

    1. Modificação dos arquivos da aplicação para que seja executada pela função principal que executa a interface do sistema, localizada no arquivo Tela_principal.py.
    2. Tarefas feitas do TO DO:
        1.  Criar arquivo de autenticação para o arquivo Database.py que não pode ser anexada ao github;
        2.  Fazer arquivos que não serão levados ao GitHub, pois contem as credenciais do Mongo e do Drive.
        3.  Fazer o merge do Sandman com o Master.
        4.  Salvar o valor da face média PSI no drive e no mongo.
    3.  Modificação da função de detecção de faces no arquivo Tela_principal.py.
    4. Criada função que salva PSI no mongo e no drive.
    5. Criação de uma função de detecção no arquivo HaarLike.py para a função das eigenfaces.

## 29/10/2020

    1.  Modificação das funções do arquivo Google_Drive.py, sendo colocados as funções para obter as credenciais do Drive, assim não é necessário abrir um sessão externa para a autenticação do mesmo.
    2.  Inserção das imagens PHI e dos pesos OMEGA no Drive e inserção dos IDs dos JSONs das imagens(Phi e Omega) no banco de dados (Mongo).
    3.  Refatoração do código de obtenção das Eigenfaces no arquivo Eigen.py.
    4.  Modificação do arquivo changelog.md, adicionando as modificações feitas na aplicação e no TO DO.
    5.  Tarefas feitas do TO DO 
        1.  Arrumar a função do Google Drive;
        2.  Arrumar a função do Banco de Dados para colocar o indexador das imagens;

## 28/10/2020

    1. Modificação do arquivo Eigen.py, gerando a função get_eigenface(), sendo esta a mais atualizada e funcional para a geração das Eigenfaces.
    2. Criação do arquivo Google_Drive.py, sendo este o arquivo que faz a conexão com o google drive para a obtenção das imagens do sistema.
    3. Modificação do arquivo AbreIMG.py, modificações na função saveImg() para salvar as Eigenfaces.
    4. Modificação do arquivo HaarLike.py, modificação na função cutFace() para o corte da face de teste, necessário para o teste dos omegas gerados no arquivo Eigen.py.
    5. Criação do arquivo changelog.md, contém as modificações feitas no projeto e as próximas modificações necessárias a serem feitas.
    6. Modificação do arquivo README.md, adicoção do nome da professora orientadora.
    7. Remoção da chave de autenticação do arquivo Database.py.

AbreIMG.py
import sys
import os
import cv2 as cv
import numpy as np

class abreImg:
    def __init__(self):
        pass
    def openImg(self,path):
        image = []
        for filePath in sorted(os.listdir(path)):
            imagePath = os.path.join(path,filePath)
            im = cv.imread(imagePath)
            im = np.float32(im)/255.0
            image.append(im)
            imFlip = cv.flip(im,1)
            image.append(imFlip)
        return image
    def openImgray(self,path):
        image = []
        for filePath in sorted(os.listdir(path)):
            imagePath = os.path.join(path,filePath)
            im = cv.imread(imagePath,cv.IMREAD_GRAYSCALE)
            im = np.float32(im)/255.0
            image.append(im)
            imFlip = cv.flip(im,1)
            image.append(imFlip)
        return image
    def createDataMatrix(self,image):
        numImg = len(image)
        sz = image[0].shape
        data = np.zeros((numImg,sz[0] * sz[1]),dtype=np.float32)
        for i in range(0, numImg):
            img = image[i].flatten()
            data[i,:] = img
        image.clear()
        return data,numImg
    def cria(self,image):
        numImg = len(image)
        sz = image[0].shape
        data = np.zeros((numImg,sz[0] * sz[1] * sz[2]),dtype=np.uint8)
        for i in range(0, numImg):
            img = image[i].flatten()
            data[i,:] = img
        image.clear()    
        return data,numImg
    def saveImg(self,img_np,name,PATH):
        try:
            os.chdir(PATH)
            # print("VALOR DO IMG_NP: "+str(img_np))
            normimg = cv.normalize(img_np, None, alpha=0, beta=1,norm_type=cv.NORM_MINMAX, dtype=cv.CV_32F)
            normimg = normimg*255.0
            # print("VALOR DO TESTE: "+str(teste))
            # teste = np.array(img_np)
            cv.imwrite(name,normimg)
        except OSError:
            print("Caminho " + str(PATH) +" inválido.")
    def openImg2DB(self,path):
        image = []
        for filePath in sorted(os.listdir(path)):
            imagePath = os.path.join(path,filePath)
            im = cv.imread(imagePath)
            im = np.float32(im)/255.0
            image.append(im)
            # imFlip = cv.flip(im,1)
            # image.append(imFlip)
        return image

Database.py
import pymongo

PATH = "C://Users//Dymytry//Desktop//TCC Alleff//git-TCC//"

class db_mongo:
    def __init__(self):
        print("TESTE")

    def conecta(self):
        # global client
        open_arq = PATH + "key.txt"
        filea = open(open_arq,"r")
        client = pymongo.MongoClient(str(filea.readline()))
        return client

    def kill_connection(self,client):
        print("CLOSING THE CONNECTION WITH DB.")
        client.close()

    def gamma2db(self,client,img_obj,tipo):
        db = client.tcc
        if tipo == "PSI":
            collection = db.mean
        elif tipo == "COV":
            collection = db.cov
        elif tipo == "PHI":
            collection = db.images
        inserted = collection.insert_one(img_obj).inserted_id
        print("Arquivo inserido com sucesso, ID:"+str(inserted))


    def get_psiM(self,client):
        db = client.tcc
        collection = db.mean
        psi = collection.find()
        return psi

    def get_covM(self,client):
        db = client.tcc
        collection = db.cov
        cov = collection.find()
        return cov
    
    def get_phiM(self,client):
        db = client.tcc
        collection = db.images
        phi = collection.find()
        return phi

    def gamma2db_mod(self,client,gamma):
        # client = db_mongo.conecta(self)
        # print(client)
        db = client.tcc
        collection = db.fotos3
        inserted = collection.insert_one(gamma).inserted_id
        print(inserted)

    def db2gamma(self):
        client = db_mongo.conecta(self)
        db = client.tcc
        collection = db.fotos2
        full_gamma = collection.find()
        return full_gamma

    def load_data(self):
        client = db_mongo.conecta(self)
        db = client.tcc
        collection = db.fotos
        for data in collection.find():
            print(data)

    def insert_data(self):
        db_mongo.conecta(self)
        db = client.tcc
        print(db)
        collection = db.fotos
        print(collection)
        nome = "Alleff"
        idade = 24
        test = {
            "name" : nome,
            "age" : idade
        }
        collection.insert_one(test).inserted_id
        client.close()

# mongo = db_mongo()
# teste = mongo.conecta()
# print(teste)

HaarLike.py
from tkinter import *
import numpy as np
from PIL import Image, ImageTk
import time
import cv2 as cv
from matplotlib import pyplot as plt
import os

from AbreIMG import abreImg as ai
from Eigen import eigenfaces as eg

#DIRA = "C://Users//Katy//Desktop//TCC Alleff//TCC_v1-3//CODES//IMAGENS//FRONTAL"

#Mudar PATH e DIRA caso seja mudado de computador, precisa ser o caminho absoluto
DIRA = "C://Users//Dymytry//Desktop//TCC Alleff//TCC Alleff//TCC_v1-3//CODES//IMAGENS//FRONTAL"
DIRA_MOD = "C://Users//Dymytry//Desktop//TCC Alleff//git-TCC//IMAGENS//TESTE3"
PATH = "C://Users//Dymytry//Desktop//TCC Alleff//TCC Alleff//TCC_v1-3//CODES//"

class harr:
    def detecta(self,frame):
        #face_cascade = cv.CascadeClassifier('C://Users//Katy//Desktop//TCC Alleff//TCC_v1-3//CODES//haarcascade_frontalface_default.xml')
        face_cascade = cv.CascadeClassifier(PATH + 'haarcascade_frontalface_default.xml')
        #eye_cascade = cv.CascadeClassifier('C://Users//Katy//Desktop//TCC Alleff//TCC_v1-3//CODES//haarcascade_eye.xml')
        eye_cascade = cv.CascadeClassifier(PATH + 'haarcascade_eye.xml')

        gray = cv.cvtColor(frame, cv.COLOR_BGR2GRAY)
        frame = cv.cvtColor(frame,cv.COLOR_RGB2BGR)

        faces = face_cascade.detectMultiScale(gray, 1.3, 5)

        if type(faces) != tuple:
            
            for (x,y,w,h) in faces:
                frame = cv.rectangle(frame,(x,y),(x+w,y+h),(255,0,0),2)
                # eg.show_img(self,cv.cvtColor(frame,cv.COLOR_RGB2BGR))
                self.roi_gray = gray[y:y+h, x:x+w]
                self.roi_color = frame[y:y+h, x:x+w]
            eyes = eye_cascade.detectMultiScale(self.roi_gray)
            #return self.roi_gray
            return frame
        else: 
            return frame
            #for (ex,ey,ew,eh) in eyes:
            #    cv.rectangle(self.roi_color,(ex,ey),(ex+ew,ey+eh),(0,255,0),2)
    def detecta2(self,frame):
        #face_cascade = cv.CascadeClassifier('C://Users//Katy//Desktop//TCC Alleff//TCC_v1-3//CODES//haarcascade_frontalface_default.xml')
        face_cascade = cv.CascadeClassifier(PATH + 'haarcascade_frontalface_default.xml')
        #eye_cascade = cv.CascadeClassifier('C://Users//Katy//Desktop//TCC Alleff//TCC_v1-3//CODES//haarcascade_eye.xml')
        eye_cascade = cv.CascadeClassifier(PATH + 'haarcascade_eye.xml')
        gray = cv.cvtColor(frame, cv.COLOR_BGR2GRAY)
        faces = face_cascade.detectMultiScale(gray, 1.3, 5)
        for (x,y,w,h) in faces:
            frame = cv.rectangle(frame,(x,y),(x+w,y+h),(255,0,0),2)
            self.roi_gray = gray[y:y+h, x:x+w]
            self.roi_color = frame[y:y+h, x:x+w]
        eyes = eye_cascade.detectMultiScale(self.roi_gray)
        return self.roi_gray
    def cutFace(self):
        image = []
        for filePath in sorted(os.listdir(DIRA_MOD)):
            imagePath = os.path.join(DIRA_MOD,filePath)
            im = cv.imread(imagePath)
            image.append(im)
            # imFlip = cv.flip(im,1)
            # image.append(imFlip)
        lgt = len(image)
        print(lgt)
        os.chdir("./IMAGENS/TESTE3_FT")
        y = 1
        for x in range(0,lgt):
            aux = image[x]
            frame = self.detecta2(aux)
            frame = cv.resize(frame, (300,300))
            cv.imwrite(str(y)+"-12.jpg",frame)
            y = y + 1
        image.clear()
    def detect4eigen(self,frame):
        face_cascade = cv.CascadeClassifier(PATH + 'haarcascade_frontalface_default.xml')
        #eye_cascade = cv.CascadeClassifier('C://Users//Katy//Desktop//TCC Alleff//TCC_v1-3//CODES//haarcascade_eye.xml')
        eye_cascade = cv.CascadeClassifier(PATH + 'haarcascade_eye.xml')
        gray = cv.cvtColor(frame, cv.COLOR_BGR2GRAY)
        faces = face_cascade.detectMultiScale(gray, 1.3, 5)
        roi_color = 0
        if type(faces) != tuple:
            for (x,y,w,h) in faces:
                frame = cv.rectangle(frame,(x,y),(x+w,y+h),(255,0,0),2)
                roi_gray = gray[y:y+h, x:x+w]
                roi_color = frame[y:y+h, x:x+w]
            eg.show_img(self,roi_color)
            tst = np.float32(roi_color)/255.0
            print(tst)
            print(tst.flatten())
            # Necessário enviar para o Eigen.py o frame em formato flatten(Gamma), para que seja descontado o valor de PSI, transformando em Phi o frame enviado
            # para assim pode encontrar a distancia euclidiana das imagens do treinamento.
            eg.get_omegaface(self,tst)
            frame = cv.cvtColor(frame,cv.COLOR_RGB2BGR)
            return frame
        else:
        # eyes = eye_cascade.detectMultiScale(roi_gray)
            frame = cv.cvtColor(frame,cv.COLOR_RGB2BGR)
            return frame
# root = harr()
# root.cutFace() 

Frequency.py
from reportlab.pdfgen import canvas

class GeraPDF:
    def __init__(self):
        pass
    def tabela(self, nome_pdf):
        try:
            pdf = canvas.Canvas('{}.pdf'.format(nome_pdf))
            x = 720
            lista = {'Rafaela': '19', 'Jose': '15', 'Maria': '22','Eduardo':'24'}
            for nome,idade in lista.items():
                x -= 20
                pdf.drawString(247,x, '{} : {}'.format(nome,idade))
            pdf.setTitle(nome_pdf)
            pdf.setFont("Helvetica-Oblique", 14)
            pdf.drawString(245,750, 'Lista de Convidados')
            pdf.setFont("Helvetica-Bold", 12)
            pdf.drawString(245,724, 'Nome e idade')
            pdf.save()
            print('{}.pdf criado com sucesso!'.format(nome_pdf))
        except:
            print("ERRO")

Google_Drive.py
from pydrive.auth import GoogleAuth
from pydrive.drive import GoogleDrive
import json
import numpy as np

class connect_drive:
    def connect():
        gauth = GoogleAuth()
        gauth.LoadCredentialsFile("credentials.json")
        return gauth
        
    def get_files():
        GA = connect_drive.connect()
        drive = GoogleDrive(GA)
        fileList = drive.ListFile({'q': "'root' in parents and trashed=false"}).GetList()
        # print(fileList)
        for files in fileList:
            if (files['title'] == "IMAGES_TCC"):
                fileID = files['id']
                break
        fileList = drive.ListFile({'q': "'"+fileID+"' in parents and trashed=false"}).GetList()
        for files in fileList:
            print(files['title'])
            if (files['title'] == "teste2.json"):
                # down_file = drive.CreateFile({'id':files['id']})
                # down_file.GetContentFile('95-11.jpg')
                cont_file = drive.CreateFile({'id':files['id']})
                info = cont_file.GetContentString()
                print(info)
                break
    def get_psi(GA):
        # GA = connect_drive.connect()
        drive = GoogleDrive(GA)
        fileList = drive.ListFile({'q': "'root' in parents and trashed=false"}).GetList()
        # print(fileList)
        for files in fileList:
            if (files['title'] == "IMAGES_TCC"):
                fileID = files['id']
                break
        fileList = drive.ListFile({'q': "'"+fileID+"' in parents and trashed=false"}).GetList()
        for files in fileList:
            # print(files['title'])
            if (files['title'] == "face_media.json"):
                # down_file = drive.CreateFile({'id':files['id']})
                # down_file.GetContentFile('95-11.jpg')
                cont_file = drive.CreateFile({'id':files['id']})
                info = cont_file.GetContentString()
                tst = json.loads(info)
                return tst
                # aux = np.array(info["PSI"])
                # info = cont_file.GetContentString()
                # print(info)
                # break
    def get_cov(GA,K):
        # GA = connect_drive.connect()
        drive = GoogleDrive(GA)
        fileList = drive.ListFile({'q': "'root' in parents and trashed=false"}).GetList()
        # print(fileList)
        for files in fileList:
            if (files['title'] == "IMAGES_TCC"):
                fileID = files['id']
                break
        fileList = drive.ListFile({'q': "'"+fileID+"' in parents and trashed=false"}).GetList()
        for files in fileList:
            # print(files['title'])
            if (files['title'] == "covariancia_"+str(K)+".json"):
                # down_file = drive.CreateFile({'id':files['id']})
                # down_file.GetContentFile('95-11.jpg')
                cont_file = drive.CreateFile({'id':files['id']})
                info = cont_file.GetContentString()
                tst = json.loads(info)
                return tst
                # aux = np.array(info["PSI"])
                # info = cont_file.GetContentString()
                # print(info)
                # break
    def get_teste(GA,id):
        drive = GoogleDrive(GA)
        c_file = drive.CreateFile({'id':id})
        data = c_file.GetContentString()
        tst = json.loads(data)
        return tst


    def get_json2img(GA,erro):
        # GA = connect_drive.connect()
        drive = GoogleDrive(GA)
        fileList = drive.ListFile({'q': "'root' in parents and trashed=false"}).GetList()
        # print(fileList)
        for files in fileList:
            if (files['title'] == "IMAGES_TCC"):
                fileID = files['id']
                break
        fileList = drive.ListFile({'q': "'"+fileID+"' in parents and trashed=false"}).GetList()
        # x = 1
        for files in fileList:
            # print(files['title'])
            # if (files['title'] == "pessoa"+str(x)+".json"):
                # down_file = drive.CreateFile({'id':files['id']})
                # down_file.GetContentFile('95-11.jpg')
            cont_file = drive.CreateFile({'id':files['id']})
            info = cont_file.GetContentString()
            tst = json.loads(info)
            print(type(tst["OMEGA"]))
            aux = np.array(tst["OMEGA"])
            print(type(aux))
            print(type(erro))
            if (erro == aux).all():
                print("OMEGA DO DRIVE: "+str(tst["OMEGA"]))
                print("NOME DO ARQUIVO: "+str(files['title']))
                return tst
                # info = cont_file.GetContentFile(files['title'])
                # print(tst["OMEGA"])
            # x += 1

    def get_credentials():
        gauth = GoogleAuth()
        gauth.LocalWebserverAuth()
        print("Arquivo credentials.json criado.")
    
    def create_img2json(name_arq,cont_arq,GA):
        # GA = connect_drive.connect()
        drive = GoogleDrive(GA)
        fileList = drive.ListFile({'q': "'root' in parents and trashed=false"}).GetList()
        
        for files in fileList:
            if (files['title'] == "IMAGES_TCC"):
                fileID = files['id']
                break
        
        fileSave = drive.CreateFile({'title': name_arq+str(".json"),'parents':[{'id':fileID}]})
        fileSave.SetContentString(str(cont_arq))
        fileSave.Upload()
        print("Arquivo criado com sucesso, ID: "+str(fileSave['id']))
        return fileSave['id']

# drive = connect_drive
# drive.connect()
# drive.get_files()

Tela_principal.py
from tkinter import *
import numpy as np
from PIL import Image, ImageTk
import time
import cv2 as cv
from matplotlib import pyplot as plt

#-----------Importações Próprias-------------------
from Database import db_mongo as db
from Frequency import GeraPDF as pdf
from HaarLike import harr as hr
from Eigen import eigenfaces as eg
#-----------Importações Próprias-------------------

WIDTH = 640
HEIGHT = 480

FACE_COUNT = 0
FACE_VECTOR = []

class Application:
    def __init__(self, window, master=None):
        FACE_COUNT = 0
        #db.conecta(self)
        #pdf.tabela(self,"teste")
        print("OI")
        # eg.classififcca(self)
        self.window = window

        self.cam = cv.VideoCapture(0)
        #self.cam = cv.VideoCapture(0,cv.CAP_DSHOW)
        self.cam.set(cv.CAP_PROP_FRAME_WIDTH,WIDTH)
        self.cam.set(cv.CAP_PROP_FRAME_HEIGHT,HEIGHT)
        #self.cam.set(cv.CAP_PROP_FORMAT, cv.COLOR_BGR2GRAY)
        frm = self.cam.read()

        self.master = Frame(master)
        self.master.pack()

        root.protocol("WM_DELETE_WINDOW",self.on_closing)

        self.cnvs_cont = Canvas(master, width=WIDTH, height=HEIGHT)
        self.cnvs_cont.pack(side=LEFT)

        self.form = Frame(master)
        self.form.pack()

        self.label_master1 = Frame(master)
        self.label_master1.pack()

        self.label_nome = Label(self.label_master1, text="Nome:",font=('Verdana','12'))
        self.label_nome.pack(side=LEFT)

        self.form_nome = Entry(self.label_master1, width=25)
        self.form_nome.insert(0,"NOME")
        self.form_nome.config(state="disabled")
        self.form_nome.pack(side=RIGHT)

        self.label_master2 = Frame(master)
        self.label_master2.pack()

        self.label_sem = Label(self.label_master2, text="Semestre:",font=('Verdana','12'))
        self.label_sem.pack(side=LEFT)

        self.form_sem = Entry(self.label_master2, width=25)
        self.form_sem.insert(0,"10º")
        self.form_sem.config(state="disabled")
        self.form_sem.pack(side=RIGHT)

        self.btn_cont1 = Frame(master)
        self.btn_cont1.pack()

        self.btn_foto = Button(self.btn_cont1)
        self.btn_foto['text'] = "Foto"
        self.btn_foto['font'] = ("Calibri","10")
        self.btn_foto['width'] = "12"
        self.btn_foto['command'] = self.tira_foto
        self.btn_foto.pack()

        self.btn_foto2 = Button(self.btn_cont1)
        self.btn_foto2['text'] = "Foto2"
        self.btn_foto2['font'] = ("Calibri","10")
        self.btn_foto2['width'] = "12"
        self.btn_foto2['command'] = self.tira_foto2
        self.btn_foto2.pack()

        self.btn_eigen = Button(self.btn_cont1)
        self.btn_eigen['text'] = "Foto2"
        self.btn_eigen['font'] = ("Calibri","10")
        self.btn_eigen['width'] = "12"
        self.btn_eigen['command'] = self.get_eigen
        self.btn_eigen.pack()

        self.tabela_master = Frame(master)
        self.tabela_master.pack()

        for a in range(0,5):
            for b in range(0,1):
                self.tabela = Entry(self.tabela_master, width=10, fg="blue", font=('Arial',16,'bold'))
                self.tabela.grid(row=a,column=b)
                self.tabela.insert(END,"")


        self.sc_cam() #Com cam descomenta | sem cam comenta

    def sc_cam(self):
        self.update()
        self.window.mainloop()

    # def update_table(self):
    #     for a in range(0,5):
    #         for b in range(0,1):
    #             self.tabela.insert()

    def tira_foto(self):
        #cam = cv.VideoCapture(0)
        #cam.set(cv.CAP_PROP_FRAME_WIDTH,300)
        #cam.set(cv.CAP_PROP_FRAME_HEIGHT,400)
        #frm = cam.read()
        #tst = cv.cvtColor(frm[1], cv.COLOR_BGR2GRAY)
        #print(tst)
        #img = cv.line(img, (0,0), (400,400), (255,0,0), 1)
        tst = self.cam.read()
        cv.imwrite("frame-" + time.strftime("%d-%m-%Y-%H-%M-%S") + ".jpg", cv.cvtColor(tst[1], cv.COLOR_BGR2GRAY))
        self.photo = ImageTk.PhotoImage(image = Image.fromarray(tst[1]))
        mx = WIDTH/2
        my = HEIGHT/2
        self.cnvs_cont.create_image(mx,my, image=self.photo, anchor=NW)
        print("BTN APERTADO")
        #cam.release()
    def tira_foto2(self):
        #img = cv.imread("C://Users//tcc//Documents//img1.jpg",0)
        frame_tst = self.cam.read()
        img = cv.cvtColor(frame_tst[1], cv.COLOR_BGR2GRAY)
        #img = cv.resize(frame_tst[1],(300,400))
        mx = WIDTH/4
        my = HEIGHT/10
        [x,y] = img.shape
        hist = np.zeros(256)
        #mask = np.array([[-1,1,-1],[1,8,1],[-1,1,-1]])
        #img2 = cv.filter2D(img,cv.CV_64F,mask) 
        for i in range(x):
            for j in range(y):
                hist[img[i][j]] = hist[img[i][j]] + 1
        
        histN = np.zeros(256)
        for x in range(np.size(hist)):
            for y in range(x):
                histN[x] = hist[y] + histN[x]

        #print(hist)
        plt.plot(hist)
        plt.plot(histN)          
        #img2 = cv.Laplacian(img,cv.CV_64F)
        #sobelx = cv.Sobel(img,cv.CV_64F,1,0,ksize=1)
        #sobely = cv.Sobel(img,cv.CV_64F,0,1,ksize=1)
        #sobel = sobelx + sobely
        #img = cv.line(img, (0,0), (400,400), (255,0,0), 1)
        self.photo = ImageTk.PhotoImage(image = Image.fromarray(img))
        self.cnvs_cont.create_image(mx,my, image=self.photo, anchor=NW)
        print("BTN APERTADO")
        plt.show()
    
    def get_eigen(self):
        print("COÉ")
            

    def update(self):
        ret, frm = self.cam.read()
        tst = cv.cvtColor(frm, cv.COLOR_BGR2GRAY) #Com cam descomenta | sem cam comenta 
        if ret :
            # frmhr = hr.detecta(self,frm)
            frmhr = hr.detect4eigen(self,frm) #Descomentar depois de arrumar a Tela
            #self.photo = ImageTk.PhotoImage(image = Image.fromarray(tst))
            self.photo = ImageTk.PhotoImage(image = Image.fromarray(frmhr))
            self.cnvs_cont.create_image(0,0, image=self.photo, anchor=NW)
        self.window.after(15, self.update)

    def on_closing(self):
        print("MORREU")
        self.cam.release()
        root.destroy()   

root = Tk()
Application(root)
root.mainloop()
